Clean up upload/publish tools, add initial tests

RyanJarv opened the pull request:
* upload.py now fails when an existing draft exists to avoid publishing the wrong version
* changed publish.py so it checks if tag is already released, asks for confirmation and doesn't upload draft when it doesn't currently exist
* moved duplicate functions to helper.py
* added initial python tests

Fixes: https://github.com/brave/browser-laptop/issues/13780

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/browser-laptop/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request [as needed](https://github.com/brave/browser-laptop/wiki/Pull-request-process).
- [ ] Request a security/privacy review [as needed](https://github.com/brave/handbook/blob/master/development/security.md#how-to-request-a-security-review). (Ask a Brave employee to help if you cannot access this document.)

## Test Plan:
Started adding tests in tools/test but not enabled in CI yet. Tested manually on fork.

## Reviewer Checklist:

- [ ] Request a security/privacy review [as needed](https://github.com/brave/handbook/blob/master/development/security.md#how-to-request-a-security-review) if one was not already requested.

Tests


- [x] Adequate test coverage exists to prevent regressions
- [x] Tests should be independent and work correctly when run individually or as a suite [ref](https://github.com/brave/browser-laptop/wiki/Code-Guidelines#test-dependencies)
- [x] New files have MPL2 license header

mbacchi commented on tools/publish_release.py:
```diff
@@ -1,76 +1,49 @@
-    tag = 'do-not-publish-me'
-  return create_release_draft(github, tag)
+  print("Found draft {}, release this? [y/n]".format(release['tag_name']))
+  if raw_input() != 'y':
```
I haven't run this myself, but this assumes the user is manually executing this script. Is that the only way its done today?

RyanJarv replied:
I thought just the upload was automated, not publishing?

RyanJarv replied:
Didn't see anything running it in jenkins at least, is it being run somewhere else?

My thinking is we do automate the publish part it will at least be gated.

bsclifton replied:
The whole process is automated actually- in jenkins.brave.com, you can view the following jobs:
- browser-laptop-build-linux
- browser-laptop-build-macos
- browser-laptop-build-win32
- browser-laptop-build-win64

Each of them runs the `./tools/cibuild.py` script which does npm install, builds binary, packages binary, and then uploads to GitHub

bsclifton replied:
The existing process works fine as-is, except the fact that tags aren't considered when searching for the build on GitHub. Instead of blindly using any draft, it should also use the tag name

RyanJarv replied:
Sorry for the delay, got that removed.

RyanJarv commented:
@mbacchi mind taking a look at this again?

bsclifton commented on package.json:
```diff
@@ -34,7 +34,7 @@
     "test": "cross-env NODE_ENV=test mocha \"test/**/*Test.js\"",
     "testsuite": "node ./tools/test.js",
-    "unittest": "cross-env NODE_ENV=test mocha \"test/unit/**/*Test.js\" --globals chrome,DOMParser,XMLSerializer",
+    "unittest": "python tools/test && cross-env NODE_ENV=test mocha \"test/unit/**/*Test.js\" --globals chrome,DOMParser,XMLSerializer",
```
Could we make a new task called `tooltest`? It would be great to isolate `unittest` to purely the JavaScript unit tests 😄

bsclifton replied:
or actually, since the output is just stdout (ex: `unittest-cov` wasn't modified), this should be great! Especially since this runs before push 👍

bsclifton commented on tools/test/test_upload.py:
```diff
@@ -0,0 +1,35 @@
+# License, v. 2.0. If a copy of the MPL was not distributed with this file,
+# You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#!/usr/bin/env python
```
This should be the first line, right?

RyanJarv replied:
Yep, thanks. Fixed this

bsclifton commented on tools/upload.py:
```diff
@@ -1,102 +1,64 @@
-  try:
-    for asset in release['assets']:
-      if asset['name'] == filename:
-        github.repos(BROWSER_LAPTOP_REPO).releases.assets(asset['id']).delete()
```
This logic is missing from the new script and is definitely important. There are times where a file gets partially uploaded and then network connection drops. When the build/upload job is re-triggered, it needs to wipe out the files on GitHub

RyanJarv replied:
Thinking here is if the file exists something went wrong (network or otherwise) and we would want manual intervention.

Maybe makes sense to add retry to the upload?

bsclifton replied:
@RyanJarv would definitely be great to have a retry built into the upload script- right now the retry is done manually... Opening up Jenkins and re-queueing the failed build type (ex: win32). When that is manually re-queued, the delete is needed (otherwise GitHub will return an error since the file exists- which may only be a partial upload)

bsclifton replied:
Here are two examples from this morning where the upload failed:
https://jenkins.brave.com/job/browser-laptop-build-win32/426/console
https://jenkins.brave.com/job/browser-laptop-build-win64/423/console

RyanJarv replied:
@bsclifton added some retry logic if you have a sec to take a look at this again

bsclifton commented:
@RyanJarv friendly reminder on the last comment I had left (needing to delete the existing files before uploading) for when you have time 😄 
https://github.com/brave/browser-laptop/pull/13781#discussion_r183622804

RyanJarv commented:
@bsclifton ty! apologize for the delay, will make sure to update this today or at least early next week

RyanJarv commented:
(copying this from above since it's kinda hidden)
Added some retry logic and a few tests for that in the last commit

RyanJarv commented:
Ok thanks, will look into these.

On 3 this is intended, open to changing the behavior but the intention here was to fail hard to prevent the possibility of 2 scripts running at the same time (or some other bug) to do something unintended silently, hopefully this would't come up much with the retry logic but if we do get to this point I was thinking manual intervention should be required. Might make sense to have a force option that deletes the draft and starts the upload over so we can expose this option in jenkins though?

RyanJarv commented:
@bsclifton the last 2 commits take care of what you mentioned before.

Fixed the tag naming, added the '--force' option to overwrite the release if it already exists and now cleaning up the partial upload before retrying.

mbacchi commented:
I'm giving this a :+1:, I haven't run it to test but I'll be porting much of this to brave-core so will open issues if/when I find them.

Nice work @RyanJarv!

RyanJarv commented:
Looking at the CI job's again I realized that I had the behavior wrong here and the drafts should be merged instead of deleted/overwritten.

https://github.com/RyanJarv/browser-laptop/blob/fix/cibuild-draft-upload/tools/test/test_upload.py#L18

@bsclifton Can you confirm this is the behavior we want?

bsclifton commented:
I'm going to pull this into 0.23.x so that we can take advantage of the more robust script 😄

bsclifton commented:
master 875b9ef2c324a392fef2b3c568103f3e89c048e7
0.23.x 0d234c91120531e823bbcf2a5196b18b80f067c4

RyanJarv commented:
Got a quick fix here for avoiding depending on the python requests module here: https://github.com/brave/browser-laptop/pull/14429
